### PR TITLE
Add arguments __repr__ in Distribution base class

### PR DESCRIPTION
**Pull request by nadavbh12**

### CODE REVIEW

File: torch/distributions/distribution.py [hunk 1]
```diff
@@ -221,4 +221,6 @@ def _validate_sample(self, value):
             raise ValueError('The value argument must be within the support')
 
     def __repr__(self):
-        return self.__class__.__name__ + '()'
+        param_names = [k for k,_ in self.arg_constraints.items()]
+        args_string = ', '.join(['{}: {}'.format(p, self.__dict__[p]) for p in param_names])
+        return self.__class__.__name__ + '(' + args_string + ')'
```
Comment (vishwakftw) [inline torch/distributions/distribution.py]: Will this not create a lot of clutter? If we had a large tensor parameterizing a distribution, the `__repr__` would fill up a lot of space. For example:
```
>>> import torch
>>> torch.distributions.Normal(torch.randn(10, 10), torch.rand(10, 10))
Normal(loc: tensor([[ 0.2669, -2.5083, -0.9248, -2.5580,  0.0581, -0.8053, -1.3089,  0.5454,
         -1.1018, -2.9030],
        [-0.7943,  1.1466, -0.4999,  0.4367,  0.1974,  0.8404,  0.1555, -0.0307,
         -0.3918,  1.6991],
        [ 0.4674, -0.3881,  0.2157,  0.2178, -0.8992,  0.6417, -1.0599, -1.2741,
          1.4784, -1.6111],
        [ 0.9133, -2.4415, -0.3185,  0.2513,  0.7126, -0.0511,  0.0191, -0.6843,
          0.5016, -0.7255],
        [-0.2401, -0.5157,  1.7942, -0.2079,  1.1840,  0.9210,  0.1123, -0.8723,
         -0.2457, -1.0446],
        [ 1.3834,  2.0407,  0.6839,  0.1849,  0.5977, -0.2929,  1.7955,  1.0106,
         -0.5795, -0.8875],
        [-0.6298, -0.2499,  1.3777, -0.0384, -0.4921,  0.7832,  1.3252, -0.8033,
          0.0562,  1.0407],
        [ 0.6389, -1.3752, -0.5781,  0.4153, -0.3506,  0.1727,  1.0859, -1.2366,
         -0.4017, -0.9681],
        [-0.5646,  0.4137, -1.7426, -0.1854, -0.1042, -1.4406, -0.2368, -2.9316,
          1.8967, -1.1922],
        [-0.9363, -0.3481, -0.5971,  2.0882, -0.8389, -0.7388,  0.4768,  0.7671,
         -0.2073, -1.3037]]), scale: tensor([[0.7092, 0.2489, 0.7965, 0.8367, 0.0876, 0.9370, 0.9043, 0.8322, 0.7560,
         0.5924],
        [0.7195, 0.3544, 0.7644, 0.8478, 0.0301, 0.5853, 0.5366, 0.7020, 0.0556,
         0.6009],
        [0.5069, 0.3518, 0.5553, 0.7847, 0.4687, 0.3424, 0.2620, 0.9680, 0.8361,
         0.2880],
        [0.2476, 0.0632, 0.9116, 0.2349, 0.7765, 0.4884, 0.2115, 0.4961, 0.8478,
         0.0358],
        [0.1228, 0.2393, 0.8527, 0.8694, 0.0185, 0.3833, 0.8530, 0.3855, 0.5278,
         0.4227],
        [0.6401, 0.1710, 0.0031, 0.7733, 0.3369, 0.1228, 0.6203, 0.0220, 0.9387,
         0.0853],
        [0.1819, 0.6805, 0.3538, 0.1556, 0.3126, 0.9230, 0.3187, 0.8747, 0.0509,
         0.3309],
        [0.5708, 0.9662, 0.4152, 0.7872, 0.7201, 0.8699, 0.2644, 0.4142, 0.7762,
         0.7114],
        [0.2799, 0.2036, 0.3563, 0.9897, 0.7181, 0.7042, 0.5987, 0.1086, 0.4707,
         0.1144],
        [0.8662, 0.6451, 0.2284, 0.9216, 0.5946, 0.8736, 0.0990, 0.5090, 0.7335,
         0.1160]]))
```
Instead, I think it would be nice to have the shapes of the parameters instead, like so:
```
>>> import torch
>>> torch.distributions.Normal(torch.randn(10, 10), torch.rand(10, 10))
Normal(loc: torch.Size([10, 10]), scale: torch.Size([10, 10]))
```

Comment (nadavbh12) [reply torch/distributions/distribution.py]: Good point. Since scalar distributions are also common, it's actually possible to do both: if it's a scalar - print the value, if it's a tensor - print the size.
What do you think?

Comment (ezyang) [reply torch/distributions/distribution.py]: I think it's reasonable to print values for scalars.

File: torch/distributions/distribution.py [hunk 2]
```diff
@@ -222,5 +222,8 @@ def _validate_sample(self, value):
 
     def __repr__(self):
         param_names = [k for k,_ in self.arg_constraints.items()]
-        args_string = ', '.join(['{}: {}'.format(p, self.__dict__[p]) for p in param_names])
+        args_string = ', '.join(['{}: {}'.format(p, self.__dict__[p]
+                            if len(self.__dict__[p].shape) == 0
```
Comment (vishwakftw) [inline torch/distributions/distribution.py]: You can use `self.__dict__[p].dim() == 0`.

File: torch/distributions/distribution.py [hunk 3]
```diff
@@ -222,5 +222,8 @@ def _validate_sample(self, value):
...
+                            else self.__dict__[p].shape)
```
Comment (vishwakftw) [inline torch/distributions/distribution.py]: `size()` would be better, I think.

**Review by ezyang — DISMISSED** [review]
see @vishwakftw's comments

**Review by vishwakftw — CHANGES_REQUESTED** [review]
Could you please fix the lint error? https://travis-ci.org/pytorch/pytorch/builds/415896037?utm_source=github_status&utm_medium=notification

**Comment by SsnL** [comment]
@vishwakftw how does this look now?

**Comment by vishwakftw** [comment]
I think it looks fine now, but Travis is still complaining.

**Review by vishwakftw — APPROVED** [review]
Looks good to me, could you add a test in `test_distributions` to make sure that the `__repr__` is called correctly?
